sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1191 @@
+operation as the runtime is able to perform optimizations based on the graph
+structure. After finalization the graph can be submitted for execution on a
+queue one or more times with reduced overhead.
+
+==== Graph State
+
+An instance of a `command_graph` object can be in one of two states:
+
+* **Modifiable** - Graph is under construction and new nodes may be added to it.
+* **Executable** - Graph topology is fixed after finalization and graph is ready to
+  be submitted for execution.
+
+A `command_graph` object is constructed in the _recording_ state and is made
+_executable_ by the user invoking `command_graph::finalize()` to create a
+new executable instance of the graph. An executable graph cannot be converted
+to a modifiable graph. After finalizing a graph in the modifiable state it is
+valid for a user to add additional nodes and finalize again to create subsequent
+executable graphs. The state of a `command_graph` object is made explicit by
+templating on state to make the class strongly typed, with the default template
+argument being `graph_state::modifiable` to reduce code verbosity on
+construction.
+
+.Graph State Diagram
+[source, mermaid]
+....
+graph LR
+    Modifiable -->|Finalize| Executable
+....
+
+==== Executable Graph Update
+
+A graph in the executable state can have each nodes inputs & outputs updated
+using the `command_graph::update()` method. This takes a graph in the
+modifiable state and updates the executable graph to use the node input &
+outputs of the modifiable graph, a technique called _Whole Graph Update_. The
+modifiable graph must have the same topology as the graph originally used to
+create the executable graphs, with the nodes added in the same order.
+
+==== Graph Member Functions
+
+Table 5. Constructor of the `command_graph` class.
+[cols="2a,a"]
+|===
+|Constructor|Description
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+command_graph(const property_list& propList = {});
+----
+|Creates a SYCL `command_graph` object in the modifiable state.
+Zero or more properties can be provided to the constructed SYCL `command_graph`
+via an instance of `property_list`.
+
+Preconditions:
+
+* This constructor is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `propList` - Optional parameter for passing properties. No `command_graph`
+  constructor properties are defined by this extension.
+
+|===
+
+Table 6. Member functions of the `command_graph` class.
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+node add(const property_list& propList = {});
+----
+|This creates an empty node which is associated to no task. Its intended use is
+either a connection point inside a graph between groups of nodes, and can
+significantly reduce the number of edges ( O(n) vs. O(n^2) ). Another use-case
+is building the structure of a graph first and adding tasks later.
```

Comment:
How can you add tasks (commands) later?  I don't see any API that allows that.

Reply:
You're right it's not supported by the current proposed API, I'll rephrase that sentence. It was supported in an earlier version of the proposal, but we don't think its basic functionality. That's why we took it out and might reconsider it being added in the future. I'll also add a note.

Reply:
This sentence about adding tasks later is still in the current version.

Reply:
Removed now https://github.com/reble/llvm/pull/114